lib/scope.js

```diff
@@ -20,40 +20,56 @@ let escape_regex = function( to_escape ) {
     * 
     * The report has a key for each scenario and each of those
     * keys should correspond to a list containing the relevant data. */
-    if ( env_vars.DEBUG ) { console.log(`Report: key: ${ key }, value: ${ value }`) }
+    if ( env_vars.DEBUG ) { console.log(`Report: type: ${ type }, value: ${ value }`) }
 
-    if ( !scope.report ) { scope.report = {}; }
-    if ( !scope.safe_id ) { scope.safe_id = `_report_${ Date.now() }` }
-    if ( !scope.report[ scope.safe_id ] ) {
-      scope.report[ scope.safe_id ] = {};
+    if ( !scope.report ) { scope.report = new Map(); }
+    if ( !scope.scenario_id ) { scope.scenario_id = `_report_${ Date.now() }` }
+    if ( !scope.report.get( scope.scenario_id )) {
+      scope.report.set( scope.scenario_id, [] )
     }
-    if ( !scope.report[ scope.safe_id ][ key ] ) {
-      scope.report[ scope.safe_id ][ key ] = [];
-    }
-    scope.report[ scope.safe_id ][ key ].push( value );
+    scope.report.get( scope.scenario_id ).push({ type, value });
 
-    return scope.report[ scope.safe_id ];
+    return scope.report[ scope.scenario_id ];
   },  // Ends scope.addToReport()
 
+  convertToOriginalStoryTableRow: async function( scope, { row_data }) {
+    /* Returns the original data of row into a string formated as a story table row.
+    * Should this show the actual row data that was used instead? That could
+    * be pretty confusing to most devs. */
+    let source_row = row_data.source || row_data;
+    let original = source_row.original;
+    let var_str = await scope.toTrimmedJSON( scope, { str: original.var });
+    let value_str = await scope.toTrimmedJSON( scope, { str: original.value });
+    let trigger_str = await scope.toTrimmedJSON( scope, { str: original.trigger });
+    return `${ ' '.repeat(6) }| ${ var_str } | ${ value_str } | ${ trigger_str } |`;
+  },  // Ends scope.convertToOriginalStoryTableRows()
+
+  toTrimmedJSON: async function( scope, { str='' }) {
+    /* Turn into JSON string then trim quotes from the start and end. */
+    let json = JSON.stringify( str );
+    let trimmed = json.replace( /^"/, '' ).replace( /"$/, '' );
+    return trimmed;
+  },  // Ends scope.toTrimmedJSON()
+
   getPrintableReport: async function( scope ) {
     /* Return a string genereated from test report data. Can/should this be extensible? */
     let report = `Assembly Line Kiln Automated Testing Report - ${ (new Date()).toUTCString() }\n\n`;
-    for ( let safe_id in scope.report ) {
+    for ( let [scenario_id, items] of scope.report ) {
 
-      report += `=== Scenario ${ safe_id } ===\nQuestion ids appeared in this order:\n`;
-      for ( let q_id of scope.report[ safe_id ].ids ) {
-        report += `- ${ q_id }\n`
-      }
+      // Might convert to markdown in the future
+      report += `---------------\nScenario: ${ scenario_id.replace(/_/g, ` `) }\n---------------\n`;
+
+      for ( let item of items ) { report += `${ item.value }\n`; }
```

Comment:
Hmm... should this print something like `${ item.type.toUpperCase() }: ${ item.value }\n`?

That way, we don't have to type `{ type: `warning`, value: `WARNING: You are missing...` every time we want to specify this is a warning etc.

Reply:
It would be nice to avoid that repetition. I think the only one that would work with that right now is `warning`, though, so I'm not sure that abstraction works well for this. We might think about instead making functionality or a function that specifically adds a warning to the report?

Reply:
#314. We've got more of these now, so it won't be so special.